recipe/2024-06-private-bicep-registries.md

```diff
@@ -0,0 +1,425 @@
+  ```
+  formData := url.Values{
+      "grant_type":   {"refresh_token"},
+      "service":      test.azurecr.io,
+      "tenant":       <tenant-id>,
+      "access_token": <aad-token>,
+      "expires_in": <5 mins in seconds> 
+    }
+    
+    // jsonResponse contains the refresh token from ACR.
+    jsonResponse, err := http.PostForm("https://test.azurecr.io/oauth2/exchange", formData)
+  ```
+- Use `RefreshToken` in ORAS auth client to authenticate private ACR.(Link to ORAS API docs for auth client : https://pkg.go.dev/oras.land/oras-go@v1.2.6/pkg/registry/remote/auth#Credential)
+  ```
+  repo.Client = &auth.Client{
+		Client: retry.DefaultClient,
+		Credential: auth.StaticCredential(test.azurecr.io, auth.Credential{
+			RefreshToken: <refresh-token>,
+		}),
+	}
+  ```
+
+**AWS**:
+
+Similar to Azure Federated Identity, users can leverage AWS IAM Roles for Service Accounts (IRSA) to authenticate seamlessly to Amazon Elastic Container Registry (ECR). Users are expected to provide `rolearn` as secret in secretStore resources referenced in the recipe config.
+
+- Enable Azure Workload Identity on Radius:
+  ```
+  rad install kubernetes --set global.awsIRSA.enabled=true
+  ```
+  or by configuring aws provider with AWS Credential kind `IRSA` using `rad init --full` command.
+- Retrieve the ECR Authorization token using the `roleARN` configured with AWS IRSA.
+- And use that as a access token with ORAS auth client
+  ```
+  repo.Client = &auth.Client{
+		Client: retry.DefaultClient,
+		Credential: auth.StaticCredential(123456789012.dkr.ecr.us-west-2.amazonaws.com, auth.Credential{
+			AccessToken: <ecr-auth-token>,
+		}),
+	}
+  ```
+
+### API design (if applicable)
+***Model changes***
+
+environment.tsp
+```diff
+@doc("Configuration for Recipes. Defines how each type of Recipe should be configured and run.")
+model RecipeConfigProperties {
+  @doc("Configuration for Terraform Recipes. Controls how Terraform plans and applies templates as part of Recipe deployment.")
+  terraform?: TerraformConfigProperties;
+
++  @doc(Configuration for Bicep Recipes. Controls how Bicep plans and applies templates as part of Recipe deployment.)
++  bicep?: BicepConfigProperties;
+
+  @doc("Environment variables injected during Terraform Recipe execution for the recipes in the environment.")
+  env?: EnvironmentVariables;
+}
+
++@doc(Configuration for Bicep Recipes. Controls how Bicep plans and applies templates as part of Recipe deployment.)
++model BicepConfigProperties {
++
++  @doc("Authentication information used to access private bicep registries, which is a map of registry hostname to secret config that contains credential information.")
++  authentication?: Record<RegistrySecretConfig>;
+}
+```
+`RegistrySecretConfig` contains registry credential information, we also want users to specify the type of credentials, e.g. username-password, federated credentials.
+
+Option 1: 
+
+Adding a `type` property to `RegistrySecretConfig`
```

Comment:
We'd usually call this `kind` to avoid overloading the term `type`. Every resource has a `type` that returns it's resource type, so we try to avoid using that field name for other purposes.